source/server/hot_restarting_base.cc

```diff
@@ -41,77 +43,171 @@ void HotRestartingBase::bindDomainSocket(uint64_t id, const std::string& role) {
   }
 }
 
-void HotRestartingBase::sendMessage(sockaddr_un& address, RpcBase& rpc) {
-  iovec iov[1];
-  iov[0].iov_base = &rpc;
-  iov[0].iov_len = rpc.length_;
+void HotRestartingBase::sendHotRestartMessage(sockaddr_un& address,
+                                              const HotRestartMessage& proto) {
+  uint64_t serialized_size = proto.ByteSizeLong();
+  uint64_t total_size = sizeof(uint64_t) + serialized_size;
+  // Fill with uint64_t 'length' followed by the serialized HotRestartMessage.
+  std::vector<uint8_t> send_buf;
+  send_buf.resize(total_size);
+  *reinterpret_cast<uint64_t*>(send_buf.data()) = serialized_size;
+  RELEASE_ASSERT(proto.SerializeToArray(send_buf.data() + 8, serialized_size),
+                 "failed to serialize a HotRestartMessage");
+
+  bool turned_on_blocking = false;
+  uint8_t* next_byte_to_send = send_buf.data();
+  uint64_t sent = 0;
+  while (sent < total_size) {
+    uint64_t cur_chunk_size =
+        MaxSendmsgSize < total_size - sent ? MaxSendmsgSize : total_size - sent;
+    iovec iov[1];
+    iov[0].iov_base = next_byte_to_send;
+    iov[0].iov_len = cur_chunk_size;
+    next_byte_to_send += cur_chunk_size;
+    sent += cur_chunk_size;
+    msghdr message;
+    memset(&message, 0, sizeof(message));
+    message.msg_name = &address;
+    message.msg_namelen = sizeof(address);
+    message.msg_iov = iov;
+    message.msg_iovlen = 1;
+
+    // Control data stuff, only relevant for the fd passing done with PassListenSocketReply.
+    uint8_t control_buffer[CMSG_SPACE(sizeof(int))];
+    if (proto.reply_case() == HotRestartMessage::kPassListenSocketReply &&
+        proto.pass_listen_socket_reply().fd() != -1) {
+      memset(control_buffer, 0, CMSG_SPACE(sizeof(int)));
+      message.msg_control = control_buffer;
+      message.msg_controllen = CMSG_SPACE(sizeof(int));
+      cmsghdr* control_message = CMSG_FIRSTHDR(&message);
+      control_message->cmsg_level = SOL_SOCKET;
+      control_message->cmsg_type = SCM_RIGHTS;
+      control_message->cmsg_len = CMSG_LEN(sizeof(int));
+      *reinterpret_cast<int*>(CMSG_DATA(control_message)) = proto.pass_listen_socket_reply().fd();
+      ASSERT(sent == total_size, "an fd passing message was too long for one sendmsg().");
+    }
 
-  msghdr message;
-  memset(&message, 0, sizeof(message));
-  message.msg_name = &address;
-  message.msg_namelen = sizeof(address);
-  message.msg_iov = iov;
-  message.msg_iovlen = 1;
-  int rc = sendmsg(my_domain_socket_, &message, 0);
-  RELEASE_ASSERT(rc != -1, "");
+    int rc = sendmsg(my_domain_socket_, &message, 0);
+    if (rc == -1 && errno == EAGAIN) {
+      // Let's deal with the possibility of a sendmsg() WOULDBLOCK by just... turning blocking on.
+      // Since the parent wants its recvmsg()s non-blocking, it would be bad to just *always* turn
```

Comment:
Since the parent wants nonblocking, and goes to the trouble of setting a callback handler thing on an fd event, it seems like it's important to not accidentally make the recvmsg()s in there blocking. A blanket make-it-blocking when sendmsg()ing could do that. (I didn't want to try reasoning about when the receive side can(not) be active.)

Reply:
I'm not sure I'm following. One side being non-blocking doesn't have any effect on the other side? In this code if you immediately do a blocking call after a non-blocking call without any epoll wait, you might as well just do a blocking call?

Reply:
Here's the sequence I'm worried about:
* I want to send a message with sendHotRestartMessage()
* The first thing sendHotRestartMessage() does is set my_domain_socket_ to be blocking
* my_domain_socket_ has a readable event; HotRestartingParent::onSocketEvent() is called to respond
* onSocketEvent calls receiveHotRestartMessage(Blocking::No).... but my_domain_socket_ is currently blocking. (Specifying Blocking::No there really just means "no change" since we assume we're keeping the socket non-blocking).

But, wait, just occurred to me: is this all only happening in the main thread, and so there's guaranteed to be no concurrency to worry about?

Reply:
Yup exactly, what you are doing here is effectively a blocking call, so I would just make it so. :)

Reply:
Nice, done.